feat: ability to test rule schema (#20503)
eslint/eslint: docs/src/integrate/nodejs-api.md

ESLint makes its best attempt at applying all fixes, but there is no guarantee that all fixes will be applied. As such, you should aim for testing each type of fix in a separate `RuleTester` test case rather than one test case to test multiple fixes. When there is a conflict between two fixes (because they apply to the same section of code) `RuleTester` applies only the first fix.
:::

### Testing invalid rule options and rule exceptions (fatal)

> You can test that invalid rule options are rejected (schema validation) and that the rule throws expected exceptions, by passing an optional `fatal` array to `RuleTester#run()`. Each fatal test case is an object with:

Reviewer: ```suggestion
You can test that invalid rule options are rejected (schema validation) and that the rule throws expected exceptions, by passing an optional `fatal` array to `RuleTester#run()`. Each fatal test case is an object with the following properties:
```

    - `message` (string or RegExp): The expected error message.
    - `name` (string): The expected error name (e.g. `"SchemaValidationError"` for schema validation failures).
- `before` (function, optional): Function to execute before testing the case.
- `after` (function, optional): Function to execute after testing the case regardless of its result.
- `filename` (string, optional): The filename for the given case (useful for rules that make assertions about filenames).
- `only` (boolean, optional): Run this case exclusively for debugging in supported test frameworks.

Any additional properties of a test case will be passed directly to the linter as config options. For example, a test case can have a `languageOptions` property.
Example: testing that the rule schema rejects invalid options and that the rule throws for a specific option:

```js
ruleTester.run("my-rule", rule, {
	valid: [{ code: "foo", options: ["allowed"] }],
	invalid: [],
	fatal: [
		{
			options: ["disallowed"],
			error: { name: "SchemaValidationError" },
		},
		{
			options: ["throw"],
			error: {
				message: "Intentional throw for testing",
			},
		},
	],
});
```

::: tip

> JSON Schema (and thus schema validation) error messages may change when upgrading ESLint or its dependencies. When you only need to assert that invalid options are rejected, prefer matching `error.name` (e.g. `"SchemaValidationError"`) instead of the full `error.message`.

Reviewer: ```suggestion
JSON Schema (and thus schema validation) error messages may change when upgrading ESLint or its dependencies. When you only need to assert that invalid options are rejected, prefer matching `error.name` (`"SchemaValidationError"`) instead of the full `error.message`.
```

Since `"SchemaValidationError"` isn't an example but the recommended way for testing schema validation errors.

ESLint makes its best attempt at applying all fixes, but there is no guarantee that all fixes will be applied. As such, you should aim for testing each type of fix in a separate `RuleTester` test case rather than one test case to test multiple fixes. When there is a conflict between two fixes (because they apply to the same section of code) `RuleTester` applies only the first fix.
:::

### Testing invalid rule options and rule exceptions (fatal)

You can test that invalid rule options are rejected (schema validation) and that the rule throws expected exceptions, by passing an optional `fatal` array to `RuleTester#run()`. Each fatal test case is an object with the following properties:

- `name` (string, optional): The name to use for the test case.

> - `code` (string, optional): The source code to run. If omitted, an empty string is used.

Reviewer: ```suggestion
- `code` (string, optional): The source code that the rule should be run on. If omitted, an empty string is used.
```

So that it doesn't sound like the code will be `eval()`-ed.

- `options` (array, optional): The options passed to the rule (use invalid options to trigger schema validation errors, or options that cause the rule to throw).
- `error` (object, required): The expected exception. Must have at least one of:
    - `message` (string or RegExp): The expected error message.
    - `name` (string): The expected error name (e.g. `"SchemaValidationError"` for schema validation failures).
- `before` (function, optional): Function to execute before testing the case.
- `after` (function, optional): Function to execute after testing the case regardless of its result.
- `filename` (string, optional): The filename for the given case (useful for rules that make assertions about filenames).
- `only` (boolean, optional): Run this case exclusively for debugging in supported test frameworks.

> Any additional properties of a test case will be passed directly to the linter as config options. For example, a test case can have a `languageOptions` property.
> Example: testing that the rule schema rejects invalid options and that the rule throws for a specific option:

Reviewer: ```suggestion
Any additional properties of a test case will be passed directly to the linter as config options. For example, a test case can have a `languageOptions` property.

Example: testing that the rule schema rejects invalid options and that the rule throws for a specific option:
```

To separate the example from the previous sentence as it isn't related to it.

```js
ruleTester.run("my-rule", rule, {
	valid: [{ code: "foo", options: ["allowed"] }],
	invalid: [],
	fatal: [
		{
			options: ["disallowed"],
			error: { name: "SchemaValidationError" },
		},
		{
			options: ["throw"],
			error: {
				message: "Intentional throw for testing",
			},
		},
	],
});
```

::: tip

> JSON Schema (and thus schema validation) error messages may change when upgrading ESLint or its dependencies. When you only need to assert that invalid options are rejected, prefer matching `error.name` (e.g. `"SchemaValidationError"`) instead of the full `error.message`.

Reviewer: ```suggestion
JSON Schema (and thus schema validation) error messages may change when upgrading ESLint or its dependencies. When you only need to assert that invalid options are rejected, prefer matching `error.name` (`"SchemaValidationError"`) instead of the full `error.message`.
```

Since `"SchemaValidationError"` isn't an example but the error name that should be used for schema validation errors.
